app/src/main/cpp/skyline/kernel/memory.cpp

```diff
@@ -7,13 +7,165 @@
 #include "types/KProcess.h"
 
 namespace skyline::kernel {
-    MemoryManager::MemoryManager(const DeviceState &state) : state(state) {}
+    MemoryManager::MemoryManager(const DeviceState &state) noexcept : state(state), setHeapSize(), chunks() {}
 
-    MemoryManager::~MemoryManager() {
+    MemoryManager::~MemoryManager() noexcept {
         if (base.valid() && !base.empty())
             munmap(reinterpret_cast<void *>(base.data()), base.size());
     }
 
+    std::map<u8 *,ChunkDescriptor>::iterator MemoryManager::upper_bound(u8 *address) {
+        std::map<u8 *,ChunkDescriptor>::iterator result{chunks.begin()};
+
+        if (chunks.size() != 1) [[likely]]
+            while (result->first <= address) {
+                ++result;
+                if (result->first + result->second.size == addressSpace.end().base())
+                    break;
+            }
+
+        return result;
+    }
+
+    void MemoryManager::MapInternal(std::pair<u8 *, ChunkDescriptor> *newDesc) {
+        // The chunk that contains / precedes the new chunk base address
+        auto firstChunkBase{upper_bound(newDesc->first)};
+        while (newDesc->first <= firstChunkBase->first)
```

Comment:
I have to anyway since i need either the previous chunk or the chunk that contains the start of the new chunk (but doesn't start at the same address as the new one). But yeah i could have used lower_bound just as easily